scipy/spatial/ckdtree/src/ckdtree_query.cxx

```diff
@@ -0,0 +1,407 @@
+        distance_upper_bound = std::pow(distance_upper_bound,p);
+
+    for(;;) {
+        if (inf->node->split_dim == -1) {
+            node = inf->node;
+
+            // brute-force
+            for (i=node->start_idx; i<node->end_idx; ++i) {
+                d = _distance_p(
+                        self->raw_data+self->raw_indices[i]*m,
+                        x,p,m,distance_upper_bound);
+                    
+                if (d < distance_upper_bound) {
+                    // replace furthest neighbor
+                    if (neighbors.n == k)
+                          neighbors.remove();
+                    neighbor.priority = -d;
+                    neighbor.contents.intdata = self->raw_indices[i];
+                    neighbors.push(neighbor);
+
+                    // adjust upper bound for efficiency
+                    if (neighbors.n == k)
+                        distance_upper_bound = -neighbors.peek().priority;
+                }
+            }
+            
+            // done with this node, get another                
+            if (q.n == 0) {
+                // no more nodes to visit
+                break;
+            } else {
+                it = q.pop();
+                inf = (nodeinfo*)(it.contents.ptrdata);
+                min_distance = it.priority;
+            }
+            
+        } else {
+            inode = inf->node;
+
+            // we don't push cells that are too far onto the queue at all,
+            // but since the distance_upper_bound decreases, we might get 
+            // here even if the cell's too far
+            if (min_distance > distance_upper_bound*epsfac) {
+                // since this is the nearest cell, we're done, bail out 
+                break;
+            }
+            // set up children for searching
+            if (x[inode->split_dim] < inode->split) {
+                near = inode->less;
+                far = inode->greater;
+            } else {
+                near = inode->greater;
+                far = inode->less;
+            }
+            // near child is at the same distance as the current node
+            // we're going here next, so no point pushing it on the queue
+            // no need to recompute the distance or the side_distances
+            inf->node = near;
+
+            // far child is further by an amount depending only
+            // on the split value; compute its distance and side_distances
+            // and push it on the queue if it's near enough
+            inf2 = nipool.allocate();
+            inf2->node = far;
+            
+            it2.contents.ptrdata = (void*) inf2;
+            
+            // most side distances unchanged
+            for (i=0; i<m; ++i) {
+                inf2->side_distances[i] = inf->side_distances[i];
+            }
+
+            // one side distance changes
+            // we can adjust the minimum distance without recomputing
+            if (NPY_LIKELY(p==2.)) {
```

Comment:
I only used ```NPY_LIKELY``` in two places to make sure the Euclidian distance case is accelerated. I will gladely remove it. It was used in more locations in the C++ parallel processing code I removed.